### PR TITLE
Add ContactInfo handling for shred versioning

### PR DESCRIPTION
**Pull request by gregcusack**
#### Problem
No support for new ClusterInfo for shred versioning/matching

#### Summary of Changes
add in ContactInfo support

Fixes #

### CODE REVIEW

File: gossip/src/cluster_info.rs [hunk 1]
```diff
@@ -3082,6 +3082,7 @@ fn filter_on_shred_version(
                 // Allow contact-infos so that shred-versions are updated.
                 CrdsData::LegacyContactInfo(_) => true,
                 CrdsData::NodeInstance(_) => true,
+                CrdsData::ContactInfo(_) => true,
```
Comment (behzadnouri) [inline gossip/src/cluster_info.rs:3085]: Can this match arm go first so it is just below the comment re contact-infos?

File: gossip/src/cluster_info.rs [hunk 2]
```diff
@@ -3091,6 +3092,7 @@ fn filter_on_shred_version(
                 // shred-version changes
                 CrdsData::LegacyContactInfo(node) => node.pubkey() == from,
                 CrdsData::NodeInstance(_) => true,
+                CrdsData::ContactInfo(node) => node.pubkey() == from,
```
Comment (behzadnouri) [inline gossip/src/cluster_info.rs:3095]: Can this match arm go first so it is just below the comment re contact-infos?

File: gossip/src/crds.rs [hunk 3]
```diff
@@ -242,6 +242,11 @@ impl Crds {
                     CrdsData::DuplicateShred(_, _) => {
                         self.duplicate_shreds.insert(value.ordinal, entry_index);
                     }
+                    CrdsData::ContactInfo(node) => {
+                        // Can't insert into self.nodes since not all validators support ContactInfo yet
+                        // Querying self.nodes (get_nodes_contact_infos()) assumes we're pulling LegactContactInfo
+                        self.shred_versions.insert(pubkey, node.shred_version());
```
Comment (behzadnouri) [inline gossip/src/crds.rs:248]: hmmm, wondering if we should hold off updating `self.shred_versions` for now until we have switched to use the new contact-info.
e.g. you may get legacy and new contact-infos from the same node with different shred-versions.

File: gossip/src/cluster_info.rs (hunk 1 above)
Comment (gregcusack) [reply gossip/src/cluster_info.rs:3085]: ya np

**Review by behzadnouri — APPROVED** [review]
lgtm, thanks.
I think we want this to be backported to v1.17 as well.

**Pull request merged**